cedar-lean/Cedar/Thm/Validation/Validator.lean

```diff
@@ -0,0 +1,401 @@
+    simp only [mapOnVars, evaluate, List.mapM₁, List.attach_def, List.mapM_pmap_subtype (fun x => evaluate x request entities)]
+    simp only [List.mapM_cons, bind_assoc, pure_bind]
+    rw [h₁]
+    simp [List.mapM_map]
+    have h₂ : ∀ (x₁ : Expr), x₁ ∈ t → EvaluatesSubst x₁ request entities :=
+    by
+      simp [h₀] at ih₁
+      obtain ⟨_, h₂⟩ := ih₁
+      exact h₂
+    rw [forall_map h₂]
+
+theorem substitute_action_nil_record : ∀ (uid : EntityUID),
+  substituteAction uid (.record []) = .record []
+:= by
+  intro uid
+  simp only [substituteAction, mapOnVars, List.attach₂, List.pmap, List.map_nil]
+
+theorem substitute_action_cons_record : ∀ (h : Attr × Expr) (t : List (Attr × Expr)) (uid : EntityUID),
+  substituteAction uid (.record (h :: t)) =
+  .record ((h.fst, substituteAction uid h.snd) :: t.map (fun (a, e) => (a, substituteAction uid e)))
+:= by
+  intro h t uid
+  simp only [substituteAction, mapOnVars, List.attach₂, List.map_pmap_subtype_snd, List.map_cons]
+
+theorem forall_map_record {t : List (Attr × Expr)} {f g : (Attr × Expr) → Result (Attr × Value)}
+(h₀ : ∀ x₁ ∈ t, f x₁ = g x₁) :
+  List.mapM (fun x => f x) t = List.mapM (fun x => g x) t
+:= by
+  induction t
+  case nil =>
+    simp only [List.mapM_nil]
+  case cons h tl ih =>
+    simp only [List.mapM_cons]
+    simp [List.mem_cons] at h₀
+    obtain ⟨h₁, h₂⟩ := h₀
+    rw [h₁, ih]
+    assumption
+
+theorem evaluates_subst_record {axs : List (Attr × Expr)} {request : Request} {entities : Entities}
+(ih₁ : ∀ axᵢ, axᵢ ∈ axs → EvaluatesSubst axᵢ.snd request entities) :
+  evaluate (substituteAction request.action (Expr.record axs)) request entities =
+  evaluate (Expr.record axs) request entities
+:= by
+  cases h₀ : axs with
+  | nil =>
+    rw [substitute_action_nil_record]
+  | cons h t =>
+    simp only [EvaluatesSubst] at ih₁
+    have h₁ := ih₁ h
+    simp only [h₀, List.mem_cons, true_or, true_implies] at h₁
+    rw [substitute_action_cons_record]
+    simp only [mapOnVars, evaluate, List.mapM₂, List.attach₂, List.mapM_pmap_subtype (fun (a, e) => bindAttr a (evaluate e request entities))]
+    simp only [bindAttr]
+    simp only [List.mapM_cons, bind_assoc, Except.bind_ok, pure_bind]
+    rw [h₁]
+    simp only [List.mapM_map]
+    have h₂ : ∀ x ∈ t,
+    (do
+      let v ← evaluate (substituteAction request.action x.snd) request entities
+      Except.ok (x.fst, v)) =
+    (do
+      let v ← evaluate x.snd request entities
+      Except.ok (x.fst, v))
+    := by
+      intro x xin
+      simp [h₀] at ih₁
+      obtain ⟨_, h₂⟩ := ih₁
+      simp [h₂ x xin]
+    rw [forall_map_record h₂]
+
+theorem substitute_action_nil_call : ∀ (uid : EntityUID) (xfn : ExtFun),
+  substituteAction uid (.call xfn []) = .call xfn [] :=
+by
+  intro uid
+  simp only [substituteAction, mapOnVars, List.attach_def, List.pmap, List.map_nil, implies_true]
+
+
+theorem substitute_action_cons_call : ∀ (h : Expr) (t : List Expr) (uid : EntityUID) (xfn : ExtFun),
```

Comment:
Naming convention:  we generally use `x` for `Expr` variables, `xs` for lists (or sets) of `Expr`s, and we reserve `h` for hypothesis names.